### ISSUE TITLE
Suggestions, Ideas & Feature Requests

### ISSUE THREAD

**Issue by Arkanic** [Anuken/Mindustry#545]
i've noticed a sudden influx of ideas, and just to clean things up, please comment them here.

**Comment by Arkanic** [comment]
Just thinking about some stuff here...

I have noticed that there is a support mech (the Tau) but what about some kind of defense mech?
something like an aegis with a "Shield project" ability (Player hits key/double taps, shield similar to the force projector appears, effectively temporarily shielding the surrounding area (assuming the player gets locked in place). On second thought, the Fortress mech could play this role, (it may make it to OP tho) seeing as it would fit with the Mech's description.

~~~Totally didn't get the idea from the SW battlefront 2 personal shield~~~

**Comment by Anuken** [comment]
Moving older issues here.

*Luxray5474:*
>What if there was an option that toggles the camera between following the unit like normal and the camera controlling the unit like in mobile? That would be useful for quickly glancing at a particular area in advance before a slow ground mech reaches it.

**Comment by Arkanic** [comment]
*Olo123pl:*
>I think about the slag and mass driver and invented a weapon that would shoot like a catapult and create a temporary slosh of slag on the path

**Comment by Arkanic** [comment]
*xslendix:*
> Here are some stuff that i wanna see in mindustry (please read it took me a lomg time to make):
> 
> * Material counter
>   
>   * When clicked, a GUI will appear where it allows you to set how many materials you want, what type of material you want, the operation (<, >, =<, =>, ==).
>   * If the condition is met, the materials are sent forward and the rest are either checked again or sent on the side (option in the GUI).
>   * The delay between checks is a minimum of 1 material(s) per second and the maximum of like idk 10 maybe?
>   * Creates power when condition is met (needs to be powered, can act like a power node i guess)
>   * This sounds like a bad idea but at least try it. Would make us able to create advanced stuff
> * LEDs. Yes.
>   
>   * The color can be chosen from a GUI
>   * Will lit up if has power
>   * Brightness is determined by the amount of power it receives
> * Material blocker
>   
>   * It can be powered (toggle or minecraft door mode, this can be changed from a GUI) or toggled by a player.
> * Electricity power level changer (change the name is more or less like a resistor)
>   
>   * When fed power, the nearest node will get that power that has been received reduced my an amount set in a GUI
> * Inverters
>   
>   * Kinda like redstone torches in mintcraft is fed power, it outputs no power, it gets no power, emits a small amount of power.
> * Power amplifier
>   
>   * Has GUI to have a slider to change hoe much to amplify the power
>   * Has 1 input: power
>   * Has 1 output: power (but wait that isnt possible to code, right? WRONG! The power amplifier can be 2x2 tile big, each tile having separate connections)
> * There are too many GUIs i know
> * I should probably sleep now
> * Help
> 
> Here are some minor changes I would make to the game:
> 
> * The energy light beams must be at least a little bit transparent (Because LEDs)
> * Be able to get an exact amount of items out of player's inventory.

**Comment by XieYongcheng** [comment]
translation:
/core/assets/bundles/bundle_zh_CN.properties: 936
```
- 终级水泵。速度是机械泵的三倍。是唯一能够回收熔岩的泵。
+ 终级水泵。速度是机械泵的三倍。
```
/core/assets/bundles/bundle_zh_TW.properties: 936
```
- 終極泵。輸出速度是機械泵的三倍並且是唯一能夠抽熔岩的泵。
+ 終極泵。輸出速度是機械泵的三倍。
```
"回收熔岩" means output slag

**Comment by Luxray5474** [comment]
I am so with the material counter, material blocker and the power level changer. I think that the second one will make research and testing way easier; it would have been nice to have that a long time ago.

**Comment by Luxray5474** [comment]
So I've just played b82 after many many finishing projects and all and here is my list of suggestions so far (will update):

**What I love so far: **

* the game overall feels..... more... ***quality***, like one of the games in this style that you pay for on steam and has existed for years. it's really amazing. all smooth, the shadows are perfect-er, the slow-down of the camera is perfect.
* The new start screen design is gorgeous.
* The new item menu icons are AWESOME. (see 4)
* I love the new outline around the turrets. (see 3)

**UI:**

1. The start screen is gorgeous, only thing is that the gray color when you hover over a selection is very bland, if there is a blur effect possible, that would be work wonders. I suggest a slight white and transparent color with a blur effect.
2. That gray outline on every UI element in game is a little weird and TBH that yellow one was much much better. 
3. The gray outline around the turrets are perfect but tbh it should be consistent with the mechs and enemies. It is debatable but an outline could look nice around the mechs and enemies.
4. The defense icon looks like a fox (aww) but not "defense" enough.

**Comment by xslendix** [comment]
Here are some stuff that i wanna see in mindustry (please read it took me a lomg time to make):
* Material counter
  * When clicked, a GUI will appear where it allows you to set how many materials you want, what type of material you want, the operation (<, >, =<, =>, ==). 
  * If the condition is met, the materials are sent forward and the rest are either checked again or sent on the side (option in the GUI). 
  * The delay between checks is a minimum of 1 material(s) per second and the maximum of like idk 10 maybe? 
  * Creates power when condition is met (needs to be powered, can act like a power node i guess)
  * This sounds like a bad idea but at least try it. Would make us able to create advanced stuff
* LEDs. Yes.
  * The color can be chosen from a GUI
  * Will lit up if has power
  * Brightness is determined by the amount of power it receives
* Material blocker
  * It can be powered (toggle or minecraft door mode, this can be changed from a GUI) or toggled by a player.
* Electricity power level changer (change the name is more or less like a resistor)
  * When fed power, the nearest node will get that power that has been received reduced my an amount set in a GUI
* Inverters
  * Kinda like redstone torches in mintcraft is fed power, it outputs no power, it gets no power, emits a small amount of power.
* Power amplifier
  * Has GUI to have a slider to change hoe much to amplify the power
  * Has 1 input: power
  * Has 1 output: power (but wait that isnt possible to code, right? WRONG! The power amplifier can be 2x2 tile big, each tile having separate connections)
* There are too many GUIs i know
* I should probably sleep now
* Help

Here are some minor changes I would make to the game:
* The energy light beams must be at least a little bit transparent (Because LEDs)
* Be able to get an exact amount of items out of player's inventory.

i was told to move it here. in the meantime, ill start working on the game and see how much i can implement.

**Comment by Arkanic** [comment]
@Luxray5474 Anuken has really made an excellent game, hasn't he?

**Comment by ZkyLB** [comment]
Ofc he has.

**Comment by Angramme** [comment]
I think a vote system would really help the game. I see it working like: someone types "/vote surrender" or "/vote kick player" in chat, then a box pops up on the left of the screen with the action, and you have to choose yes or no. However if someone types for example "/vote map" there are multiple options to choose from like the names of the maps... I know from experience that it takes time to implement such feature, but IMHO it would really help the experience

**Comment by Arkanic** [comment]
@Angramme that's a good idea

**Comment by Angramme** [comment]
@Arkanic thanks :)

**Comment by ataricom** [comment]
Here's a wish list of things I'd love to see in the game. 
- Custom pathfinding nodes for friendly mechs. They only take the shortest and most direct path but other players can easily exploit that by setting up defenses that mechs are guaranteed to walk right into them.
- Switches for quick enabling/disabling power or a more refined power node connection interface.
- Sorters that can allow the chosen material to pass straight through rather than out one of the sides. 
- Palette change for the laser drill or air blast drill. I'm not sure why laser drills are no longer blue but it's really difficult to tell the two apart in the UI since they're almost the exact same shade of orange.
- With the exception of thorium I think every material should be obtainable in every map, either through direct mining or combination. If you have access to oil and have silicon in your inventory then you can make just about anything, but you're stuck if you don't have silicon.
- Following the last point, you should be able to recover scrap from destroyed enemies and enemy buildings. If you can't collect it directly then maybe one of the drones could.
- Using slag to "fuel" thermal generators.
- Fog of war for unexplored areas. 
 
I'm honestly amazed at how GOOD this game is!

**Comment by Luxray5474** [comment]
> Switches for quick enabling/disabling power or a more refined power node connection interface.

Definitely. 

> Using slag to "fuel" thermal generators.

But would that just be counterproductive and OP? You melt the scrap to make slag, then you use it for thermal generation. Energy in, energy out. That means you have a *very* unlimited source of fuel that can be exploited from enemies, though I might be wrong about that.

> Fog of war for unexplored areas.

The fog of war thing was scrapped a while back. I don't know exactly why, but I think you can look back in the Discord or something.

**Comment by Arkanic** [comment]
@Luxray5474 i agree that we shouldn't create any "infinite energy" things
~~I should probably join  the discord~~

**Comment by Spicy0903** [comment]
@Angramme I know that that was a mod for mindustry.kr on 3.5 its still their i think

**Comment by Arkanic** [comment]
@Spicy0903 the whole concept of fog of war was ditched a while ago, i think.

**Comment by Spicy0903** [comment]
@Arkanic i think you mentioned the wrong person

**Comment by TheChosenEvilOne** [comment]
@Angramme 
I made a server sided mod for exactly this in 3.5, called votemod (very original name I know)

**Comment by Arkanic** [comment]
@Spicy0903 no

**Comment by Spicy0903** [comment]
oh ok because i never mentioned fog of war i was talking about vote mod

**Comment by Arkanic** [comment]
@Spicy0903 oops
i was talking to the wrong person

**Comment by Angramme** [comment]
@TheChosenEvilOne cool! thanks! I don't run my own server but it will be surely useful for others. Meaby try to make a pull request to the server code? It should be a default feature at least IMHO

**Comment by AgentAppl** [comment]
I think it would be nice to have an area where you can see how much energy you’re producing, how much you’re using, and how many items of each material you are receiving a second, primarily the energy because I have problems sometimes seeing if I am making enough energy for everything to work. The item part would just be a bonus.

**Comment by Arkanic** [comment]
@AgentAppl that's be useful, but you may need a tab for every different network.
other than that i think it'll be pretty easy to implement.
:D

**Comment by mmm4m5m** [comment]
Hi :)
I play with 10" tablet. Probably it is a bit harder with phones and smaller displays.

1) Sometimes I do mistake and I get 2 resources together on the same conveyor. Fixing the problem is not easy and takes a lot of time.
I would like if game completely reject mixing the resources by default and resources just stop like it is a dead end. Instead there could be new structure/building for that -  conveyor mixer / conveyor funnel / conveyor mixing funnel.

2) When I build conveyors, before confirming the action ("check" button), the game accept the first conveyor that I placed. 
In case of left/right turns, it will be more easy if game override the previous conveyor and accept the last placed conveyor.

3) My "graphics" settings are low - 5 FPS, Display Effects NO, Pixelate YES. I still feel like there are too many details. Even with maximal zoom out. 
I would like more settings and more simplified graphics. Pixelate it more. One map square could be displayed with 1 color, (like it is 1 pixel). Or display one map square with 4 colors, (like it is 4 pixels, 2x2).   Allow lower FPS, like FPS 1.

4) Swipe minimap will change minimap zoom. Tap minimap will show bigger minimap. Swipe the big-minimap will change minimap zoom.
I would like if tap minimap move the view, move the map to that point. 
Swipe the big-minimap is enough for minimap zoom. Therefore swipe the minimap could move the whole view.

5) Generally, I feel it is hard to zoom in/out, scroll the map and look around my base (the core). Zooming in and out is not smooth. 
I would like overlay buttons for zoom in and zoom out.

6) I would like to watch the enemies but my main worker / main ship will follow me everywhere and will shoot the enemies. 
I would like toolbar button: "attack / stay away"

7) Sometimes my defense is fully build and I just wait to collect more resources.
I would like toolbar button: "Fast Forward", "Fast speed", "x1 / x2 / x4 / x8". Also it will safe phone/table battery life.

8) The "close up" toolbar button (up arrow / ^) will hide both - next wave with the timer and the toolbox with all structures/buildings.
I would like if I always see the timer for the next wave.

9) This is a minor note. It was a little hard for me to understand how to build and use bridge, how to use generator and power nodes. I did not read the help. it is not 100% intuitive. But it is not hard, it is 95% intuitive.

I hope my explanation is good.
Thanks! 
Nice game!

**Comment by Daniel-Chin** [comment]
@mmm4m5m Nice thoughts! (6) makes a lot of sense. Although, personally I disagree with (1). Designing conveyer bell systems and make sure nothing gets mixed is half the fun. It's like solving puzzles. However, if a considerable portion of the players do not like this puzzle, I would recommend an "assistance mode" in gameplay settings that locks conveyer belts to single items. (Maybe that will be computationally slow, but there may be an algo for that.)  
For (7), Game designers should also be very careful with fast-forwarding time. The risk is that the player may choose to do the less fun thing to get more resources. "given the opportunity players will optimize the fun out of a game." Ask yourself: is it more fun to (A) watch accelerated waves fall under your defence, or (B) spend time improving your production while you know the time you spend is concurrent with defending enemies so the time is not wasted? In other words, I feel "not being able to fast-forward" gives me an excuse ("there's time anyway") to keep optimizing my production to higher and higher levels.  
(8) makes a lot of sense. How to design the UI so you can choose whether to hide "next wave" is the big challenge.  
Feel free to let me know what you think!  
Nice suggestions!

**Comment by Daniel-Chin** [comment]
@ataricom 
> With the exception of thorium I think every material should be obtainable in every map, either through direct mining or combination. If you have access to oil and have silicon in your inventory then you can make just about anything, but you're stuck if you don't have silicon.

For me, it was kinda fun to go past wave 15 and re-launch into a no-silicon map with silicon launched with me. The same thing goes for titanium. To get Titanium from Scrape, you need Titanium in your inventory. If a map has scrape but no Titanium, "play past wave 15 and bring your Titanium to the map" is quite fun. It offers a reason to re-launch with cuestomized starting resources.

**Comment by timo** [comment]
Please put more details about blocks in the core database; especially the materials needed to build something is an unfortunate omission

**Comment by Quiark** [comment]
It would be great to have right-angle conveyor belt routing around buildings rather than straight(-ish) line through buildings which is useless most of the time...

**Comment by timo** [comment]
i'd like there to be a faster-to-navigate block selection menu; maybe a nested radial menu that pops up when you hold a button would be good

**Comment by Arkanic** [comment]
@timo adding more descriptions seems to be my sole purpose on this repo, is there any particular descriptions you want?

**Comment by timo** [comment]
@Arkanic really just having the same information that's available in game with the "?" button would be great in the core database. for example, if i want to figure out exactly how much of what material i'll have to launch with to build a certain starting building.

**Comment by milochristiansen** [comment]
Currently construction drones are of limited use at best, it would be really nice if they could replace stuff that was destroyed automatically. This would give them a purpose, make them something well worth building.

Basically the intended use would be fixing minor battle damage, such as random destroyed conveyors when an aircraft manages the sneak through. Could be balanced by making them build quite slowly when not working on something placed directly by the player.

**Comment by Vortetty** [comment]
just an idea here, but you have added a server plugin api, could you possibly add a client modding api, that will disable mods and reload assets upon joining a server allowing modding to become easier, and prevent it from breaking servers. this would also allow people to use custom textures, etc. this level of customization might bring more attention to the game, bring more players, and possibly more revenue/help with programming

**Comment by zmxefy** [comment]
如果能有新的专门针对防空的大型炮塔

**Comment by Vortetty** [comment]
> 如果能有新的专门针对防空的大型炮塔
Translation: If there are new large turrets dedicated to air defense

**Comment by Hodhandr** [comment]
When configuring loadout, you can use the buttons to add or subtract 50 of a resource at a time. That works fine for bring a couple hundred of a resource, but becomes really annoying when I want to bring 1000+ of several resources.

Being able to either manually write in a number or use a modifier (shift, control) to add more at a time (for example, 500, 10x as much) should alleviate this.

**Comment by Arkanic** [comment]
@Hodhandr yes, this is a bit of a problem sometimes

**Comment by Tazshelby2016** [comment]
The ability to ignore certain teams and leech off of others structures needs to be added in-game. This will allow for many things including:
- Trade and bartering
- Alliances
- Espionage
- Thievery
- Economies
- Greater focus on what resources the terrain you have has
- Multiple new game modes
- Diplomacy

**Comment by maximmasterr** [comment]
#664

Java require compilation, js not. Yes, it is a lot of work, but it will be very useful, if using mods we will be able to create blocks, mechanism, game mechanics it will give infinity features. If we will use it everywhere (server, client) it will be easier. Also I think we need to reopen it because it is large theme.

**Comment by Vortetty** [comment]
> #664
> 
> Java require compilation, js not. Yes, it is a lot of work, but it will be very useful, if using mods we will be able to create blocks, mechanism, game mechanics it will give infinity features. If we will use it everywhere (server, client) it will be easier. Also I think we need to reopen it because it is large theme.

actually, creating new things with java is already very easy, you just have to define the block and the parameters. you just need basic java knowledge, a java api would be nice though.

**Comment by maximmasterr** [comment]
Ok, I am talking about plugins for every platform, it also will be good to create mode storage where you can easily install plugins

**Comment by milochristiansen** [comment]
> javascript is similar to java

Whoa there. They look superficially similar, but the two languages could not be more different if they tried. You don't "compile js to java" either.

Anyway, you can't wave a magic wand and suddenly have a working mod API that allows some kind of game extending plugins that allow you to add new game mechanics. Every single thing you want to be able to do with scripts needs special code written to provide an interface for the script to do that thing. Basically, if the game was not designed from the beginning to be scripted, adding it later is a *nightmare*.

**Comment by maximmasterr** [comment]
Mindustry has already plugins api, but they runs only on server

**Comment by mmis1000** [comment]
Can we probably separate the run wave permission from admin permission?

The sandbox mode isn't very useful in server mode without admin permission because only admin can start a new wave(I checked the code and found that they are bound altogether under a `isAdmin()` check).

It would be great if there is a option to simply allow anyone to run new wave

**Comment by Anuken** [comment]
@mmis1000 

I don't see an easy way to implement permissions like that. You would either need to implement a fine-grained customizable permission system for players, or add a new "wave runner" player data flag, which would result in a lot of copy pasted and redundant code (not something I would accept). Keep in mind that it would need support for:
- listing all people with that permission (console)
- giving people that permission (console)
- storing that permission (console) (difficult, considering the saving system is rigid and doesn't support new properties)
- revoking the permission (console)
- making sure people keep it when changing IPs and making sure it's not spoofable
- exposing it to clients (or at the very least setting and removing it by default)

I honestly don't see why you would want some un-trusted individual to schedule waves for sandbox. If you just want to play sandbox and test enemies sometimes, why not edit the map rules to have infrequent spawns instead?

**Comment by KitsuneIDK** [comment]
Please add SRV Support in the game to work with domains without need a port

**Comment by SuberFu** [comment]
The goal of this suggestion is to implement a more dynamic enemy for attack type map without having to get a complex AI to build a base.

The basic gist is to create a new enemy only buildings called "Grey Goo" and "Grey Goo Pod". "Grey Goo Pod" essentially acts as enemy's base while "Grey Goo" acts a bit like creeps (think StarCraft Zerg creeps).
"Grey Goo" is walkable (essentially like conveyor in path-ing), provides a small amount of energy, acts like a global storage vault and have a decent amount of health, very slowly heal over time, and inflict small damage to any player units on it or right next to it (including player building).
"Grey Goo" not connected to "Grey Goo Pod" will slowly "die" over time.

Enemy waves now split up into phases.
Phase 1: Dormant - Nothing happens.
Phase 2: Spread - Each "Grey Goo" block has a chance to spawn another grey goo block next to it periodically over the Spread phase. 
Phase 3: Build - the AI will randomly build building on grey goo blocks. With below logic
Defense buildings: preferentially near the "front line" (where Grey Goo have room to spread to but have not yet), with limit on how close together the defensive buildings are (AI won't spawn tight clusters of defensive buildings). "Grey Goo" will slowly supply defensive buildings.
Unit Spawn Building: the AI will spawn these buildings away from the "front line", with requirement that AI can only spawn 1 building per X amount of Grey Goo tiles available. These building will slowly spawn enemies that attack your base.
Phase 4: Assault: The "Grey Goo Pod" will spawn an attack wave, The strength of the wave is dependent on the scaled roughly as followed.
(Wave #) / sqrt(Grey Goo tiles).
The idea is that the Grey Goo Pod grows strength over time, but when it's spread out, it expends energy to maintain the Grey Goo and launch a strong wave. However, as you push back the Grey Goo toward the pod, the assault wave grows stronger.

The idea being that it's a more of less mindless enemy that slowly encroach on your base, and with attack wave growing stronger as you fight your way back against it.

**Comment by Arkanic** [comment]
interesting

**Comment by zicklag** [comment]
Me and @tekhnaeraav have just had an idea for a new kind of drone pad that allows relatively slow but long range transportation of materials using drones.

The idea is that you can create a transportation drone pad that will build drones that are similar to miner drones, but instead of mining materials, transportation drones will transport any materials loaded into their pad to any other block that can take materials such as a turret, a container, or the core. The pad would produce 2 drones per pad.

Transportation drones are similar in some ways to the function of mass drivers, but they have several different aspects that would effect your choice to use one or the other:

&nbsp; | Mass Driver | Transportation Drone
----  | ----- | -----
**Throughput** | High throughput from point A to point B | Throughput depends on distance between points and drones move relatively slowly.
**Distance** | Mass drivers must be within range of eachother. | Drones can bring material any distance across the map
**Power Consumption** | Mass drivers take 105 power/second | Drone pads will proabably take 66 power/second like miner drone pads
**Cost** | More expensive — Lead: 125, Titanium: 125, Thorium: 50, Silicon: 75 | Less expensive — ( speculated ) Lead: 55, Metaglass: 45, Silicon: 75, Titanium: 45
**Reliability** | Reliable until pad ( 360 health ) is destroyed | Reliable only if drones and pad remain in-tact. ( drones will be rebuilt if pad has power ) — Pad life: 160, Drone life: 60
**Surface Area** | 3x3 block — Easy to load and unload, but take extra space. | 2x2 — Smaller, but not as much surface to load. 
**Diversity** | Can handle any number of different materials efficiently. | Can handle any number of different materials, but individual drones can only transport one kind of material at a time. 
**Distribution** | Mass drivers unload only into another mass driver | Drones can unload into any block that accepts materials: containers, the core, turrets, routers, etc.
**Capacity** | 120 items | 75 items

### Distribution

If multiple different materials are being loaded into a transportation drone's pad, the the drones, which can only transport one material at a time, would have to have some logic used to determine which material they grab.

They could grab the material that there is the most of in the pad, but that could definitely lead to drones only ever transporting one material from a pad that has multiple materials in it. This could just be part of the way that the pads are disadvantaged from alternatives like the mass driver.

Another option is, when there are two drones active, one drone will take the most abundant resource in the pad and the other will take the second-most abundant resource in the pad.

**Comment by Arkanic** [comment]
what about a day/night cycle?

*gunfire flashing in the distance*
*factories glowing as they consume fuel*

**Comment by 1RedOne** [comment]
Developers like me would love to help!  It would be cool if there were some issues opened with the tag `Help Wanted` so we would know that there are items the devs want but haven't prioritized.

For my own selfish idea, I want a 'kill counter' to track how many units and buildings a turret has killed.  It would be extra cool if the unit could level up and gain new abilities if it survived so many waves or hit 50, 100, 200 kills, etc.

**Comment by Vortetty** [comment]
> > javascript is similar to java
> 
> Whoa there. They look superficially similar, but the two languages could not be more different if they tried. You don't "compile js to java" either.
> 
> Anyway, you can't wave a magic wand and suddenly have a working mod API that allows some kind of game extending plugins that allow you to add new game mechanics. Every single thing you want to be able to do with scripts needs special code written to provide an interface for the script to do that thing. Basically, if the game was not designed from the beginning to be scripted, adding it later is a _nightmare_.

I understand that js is not actually that similar to java. Learning it takes a little more time, and under the hood it has almost no similarities. But fundamentally if you are good at JavaScript, much of how it is written is the same. But my point was they are similar to wire so you might as well learn both.

**Comment by Anuken** [comment]
@1RedOne I currently use Trello for planned features, so most things that may require help can be [found there](https://trello.com/b/aE2tcUwF/mindustry-trello), although I do not have a 'help wanted' tag yet.

**Comment by BeanHeaDza** [comment]
An either or suggestion:

- Make the Unloader directional so that it only outputs to one direction and pulls from the other 3 sides
- Leave the current unloader as is and add a new directional unloader that costs more and requires power to operate

This would allow us to chain production buildings next to each other with only one gap for the unloader. It might be a bit too strong if you're allowed to unload any resource, so it might help making the Unloader operate when there is actually a unloader filter set.

**Comment by maximmasterr** [comment]
Anti air turret that efficient destroy air enemies or air magnets that makes air enemies attack it, not other blocks

**Comment by 1RedOne** [comment]
Should anti-air units be able to impact the ground?  I've seen enemy
Scatter units attack my character while I'm building, and they're able to
destroy new turrets under construction beneath me, or any other blocks I
happen to be above, like conveyor belts.

**Comment by maximmasterr** [comment]
Only shoot to side of drone, main thing is that turrent must use power and have long range

**Comment by Lexcess** [comment]
Hey great game, I just started playing on Android + PC, being able to transfer saves is really cool. Here are a few things I noticed while playing:

1. **Smooth out research curve** 
     - Too many items are locked behind Titanium, a lot of the tech tree can be unlocked on the first or second maps leaving a fairly long drought of unlocks till you get to the first Titanium maps
      - Notably the tier 1 fluid buildings would be good to arrive together, not being able to properly route liquids isn't much fun (e.g. get at least routers and junctions).
3. **Mechs unlock too early** 
     - you really don't need them (at least early on) and bloat out the early game tech tree. I wouldn't necessarily change the build costs, but put them behind a higher tier resource to unlock
     - Also perhaps offer a earlier flier upgrade, doesn't have to be a big upgrade, but would be a nice for variety, given that it is often the vehicle you use the most
4. **Movement controls** 
     - Android - touch controls are really great: I like to pause fly about and plan then build
     - PC - I would like it if when paused you could effectively control a ghost ship so you can move around and plan like the Android version, when you unpause either move the ship over to you (like on android) or snap you back to the ship
5. **Confirm/cancel** 
    - the behaviour of how cancelling plans feels clunky, there is nothing worse than planning out a whole section only to accidentally close a panel and lose it all and building something you don't want due to a misplace but you don't want to risk cancelling everything is annoying
    - It might be a bandaid, but an undo would help with a lot of these scenarios cancelled everything: undo, placed one thing wrong: undo
    - I still think it would be worth revisiting the planning/cancelling in general - it is hard to describe how it should be better but it feels clunky, certainly on Android it feels like if I tick confirm that shouldn't be lost when cancelling (perhaps have an x to replace the cross to unconfirm?)
6. **Level objectives/unlocks** 
    - It might be nice to see what unlocks are left on a level (i.e. wave targets to unlock later maps, loadout unlock)
    - If you do this it might be nice to have specific map challenges (certain amount of resources launched, creation of resources not naturally available on the map) maybe give people a star or some form of cosmetic award (achievement/medal/whatever)
7. **Level styles**
    - I really like the variety of maps in the game (low resource, high volume of weak enemies vs high resource. powerful enemies), here are some other ideas
    - Start with tar pits on fire, needing wave turrets to fight them back and explore further
    - Lock some higher tier resources in containers in enemy bases
    - Electrical storms that disable electrics (and fliers?) temporarily
    - Meteor showers doing heavy damage to areas without force projectors (maybe start with one for free around core?)
    - Glacier - have higher tier resources under ice that need to be chipped or melted out in some way
    - Extreme cold - melters, slag pipes, incinerators, reactors, generators etc... all create heat zones, everything outside of that area takes damage over time
    - Acid fog- everything takes light damage over time (except fliers.... above the fog?), but electricity is free (either lots or thermal or acid pools acting like infinite batteries) meaning you need menders covering all your buildings constantly

That's all for now, thanks for all the work you've put into this!

**Comment by jakiki6** [comment]
Could you add that you can choose the team when you join a PvP server.

(Sorry for my English)

**Comment by Damnae** [comment]
> the behaviour of cancelling plans feels clunky, there is nothing worse than planning out a whole section only to accidentally close a panel and lose it all and building something you don't want due to a misplace but you don't want to risk cancelling everything is annoying

Yes, right click shouldn't cancel everything you placed. Chances are you only want to remove the last thing you (mis)placed, and if it isn't the case you could use rectangle selection to cancel things anyways.

**Comment by Valink16** [comment]
A blueprint system would be a nice addon, being able to quickly place defenses and basic infrastructures saves a lot of time imo

**Comment by maxZ-2020** [comment]
A Programmable AI System.

... this is a bit crazy... but.....

If you are interessted, next year I actualy would have some time to help developing such a monster...

Required components:
    * a bus like network system which publishes values to everyone subscribed to it (think factorio circuit network)
    * all entities can publish messages if connected (broadcast, unicast)
    * all entities can receive messages and change their state according to it
    * wires connect static entites (similar to the power grid)
    * wifi adapters, connect moving entities (long range variants consume power)
    * scanners: (different kinds: short and long range) which reveal information about the environment (map tiles, ores, static entites, moving entites, everything in a given range)
    * network routers: filter and relay messages between seperate networks
    * cpu
            * a special static entity which can run some form of assembly (maybe webasembly [very secure] or lua or something)
            * the scripts read and write messages, they can control everything connected to the bus through that
            * running script can be selected via message on the bus
            * depending on MK level consumes no power or a lot, has more or less available memory and instructions per cycle
    > the player ship has a wifi built in, the cpu as well, allowing for ai control from the beginning of the game

The following cool scenarios would be possible:

Game Starts. You build the MK1 Version of the CPU. You build a construction drone (Same capabilities as the player ship)
You fly to a nearby resource field and put a Scanner down and wire it with the CPU. You send a message (playership needs to be in wifi range) to the CPU which runs the build outpost script. The script uses the scanner information and sends build orders to the construction drone.
The drone places miners in the optimal pattern and lays down belts towards the core.

You are a while into the game. You select a micro management script. The CPU uses information from the via wifi connected bots to micro the shit out of every unit. They spread out in a perfect pattern and single out the most relevant targets.

You could write base building scripts, which would build and defend a base + bots in a given area. Very interesting with enormous map sizes.

CPUs could sit ontop of movable entites, like a subcommander when you are fare away from the main base. You would have to manage power and stuff.
Wifi Jammers would make your life difficult, since your units may not stray too far away from your commander.

Players could have a dozen of scripts, that work together to play the game entirely on its own. So you could watch oposing scripts to battle it out.
Tweak those scripts and try again. Until noone can beat it anymore.

The scripts could actually be neuronal nets (webassembly is cool) trained by someone.

Hacking of enemy circuit networks. (read wifi traffic, inject false messages, convert cpus to your faction)

**Comment by Saucistophe** [comment]
A few electricity nodes could drastically improve endgame.

- Diodes (nodes that send power one way, not the other)
- NOT gates (two pins A and B generates 1 power on B, except if A is powered)
- Transistors (3 pins A, B, C: if A is powered, power can flow from A to C)

With only these three, one can address the following cases:
* Create a separate emergency power source for reactors (and cooling systems :) )
* Shut a production branch when a power plant fails
* Enable temporary rerouting of energy to defense

This would also pave the way to more computing thingies: detectors, triggers, clocks, powered doors, ...

**Comment by mmm4m5m** [comment]
No pulverizer. When I build a pulverizer then overflow gate is created.
Also getting sand is harder than getting oil.

**Comment by Arkanic** [comment]
@maxZ-2020 yes. this is what we need. also considering the fact that most of the people who play this game have at least a basic understanding of computer programming.

**Comment by Prezombie** [comment]
Factorio burnout refugee here, absolutely loving this game so far.

Campaign suggestions:

Instead of unlocking full core customization at wave 15 (or wherever it's set for higher levels), Give each map a Core Level. Each core level would increase the storage capacity, and allow for dropping of higher tier resources. Could also increase the base stats of the spirit drone and it's morphed versions.

On the loadout customization, just show all the resources available and sorted by tier, there's no reason to need to add them one at a time to a list that allows current items to have a quantity of zero just fine.

The enemy radar has no clear center point if you're moving, a faint gray overlay circle would be an easy quick fix.

Make it clearer when power is passing between touching blocks, some do it, some don't. A thin pulsing border between them would be enough.

On the selection window when making factories, showing production requirements (including power) and results would be useful, so one doesn't need to keep checking the detailed versions where they haven't memorized things yet. On the detailed page, adding resource consumption/production per second in parens would be a nice easy feature to make the math easier for nerds who want to know ideal ratios. 

Make buildings only ever output to plain conveyor belts or other buildings. It's incredibly adhoc to put that warning on the router to not plug into a resource producer directly, and just increases factory sizes for no reason.

Water pumps should need a unit of metaglass, currently you get them way too early and it can confuse a player. At the very least, Needing a little metaglass for the pump unlock (and the other water-reliant techs) prevents hitting a wall where you have the outputs ready but not the input.

Because of the chaotic nature of the shape of ore patches, Miners that are touching and gathering the same resource should be considered a single multi-block entity. Currently, making a mine maximally efficient is difficult if the ore patch is touching a wall or is large enough that the center can't be mined without putting conveyor on ore patch, and since production yield is based on exactly how many ore patch tiles the miner is covering, it's easy to overestimate how much yield you're getting. Plus, surrounding even a simple mine with belts is a hassle in a game with a high time pressure. If the game treated the miners as a multiblock entity, you could just cover cleanly, look at the total production rate, and add the correct number of belts to ship the lot.

Customizable Core/Spirit Drone colorization. 

Make the map popup show the whole map, as well as the drop point.

**Comment by Prezombie** [comment]
Instead of naively connecting two points with belts with control click+drag which totally ignored any obstacles to building, a simple A* algorithm to connect the points with the shortest path that isn't blocked by the landscape or other buildings, going to the point nearest if there's no valid simple path. 

Similarly, Clicking and dragging power nodes should be like factorio power pole placement, only placing one every n tiles.

**Comment by ducttapecoder-vt** [comment]
Sorry if any of these are repeats:
-Fluid diversity: When water and cryofluid both give the same benefit, why would anyone use cryofluid in a building? Make cryofluid more powerful in some applications.
-Key binding: allow a second key to be bound to an action. I'd like both mousewheel and R to rotate objects for example.
-Invert filter: allow all blocks _except_ the one you choose to go forwards.
-Fighter mechs should automatically attack all targets on map when given attack order. Currently, anti-air fighter mechs will head to the enemy spawn point and sit there until an enemy comes close to them. The repair drones fly across the map to get repairs done, why wouldn't your defense drones do the same?
-When importing data from standalone game into steam game, corresponding achievements should unlock. I had to start a new game data to redo the tutorial, router block, and titanium unlocks. Not a huge deal but annoying.
-Enable public game option should be a yes/no versus ok/cancel dialogue. I was afraid cancel would cancel the multiplayer game and clicked ok but then my game was open to public.
-Campaign Story? Are the spores the enemy and have taken control of your own mechs? Are you fighting a different mech army? etc.
-Diagonal conveyors should wrap around wall blocks rather than just be blocked by them. I spend a ton of time in game getting conveyors to follow walls.
-Core resource display on screen at all times. Enable by option maybe?
-Router block contents display by hover. Enable by option?
-Show junction contents when clicked (or hover)
-Click and hold resource in block display to hoover. I spend a lot of time cleaning up 'messes' in my router/junction blocks. I usually end up deleting the whole block and placing it back because it's easier than trying to remove the resources from them. If I could click and hold to grab items from a block as they come in, rather than having to continuously click, it would be more viable.
-Make it more obvious whether deleting a block gives you resources back. My group of friends still doesn't agree whether it loses or returns some/all resources.
-Make shock tower connections more obvious? I have a really hard time getting the large power towers to work correctly. It always seems like they refuse to connect. Someone usually points out that all the connections are taken, but it's not super obvious to me. Maybe a pop-up or sound to indicate that one of the two blocks has reached connection limit?
-Make something that sprays environmental effects? There's an achievement for water and electricity on an enemy but getting enemies wet is difficult.
-Molten slag in a scorch turret seems like it ought to be a thing.
-Force projectors seem to shutdown without reason. Presumably due to damage or overheat but no way to know. Show a damage indicator? Also, cryofluid should make them last longer.
-Enemy health bars should be an option. On maps with multiple bosses, it's impossible to tell the health of each boss and damage to individual bosses seems to not reflect the total health indicator in expected ways.
-On some maps there are ground decorations that look like coal and I end up accidentally mining sand instead, quite annoying. Make them more obviously different.
-We've had some pretty serious sync, display issues. To one player, a belt will look full, and to another it's half empty. We had one player host and everyone else ended up with pings of over 200000ms due to poor connection and it became difficult to tell what actions on screen were 'real' or 'fake'. Usually we'd determine an enemy was 'taking too long to die' and have to look up at the ping indicator to figure out the problem.

**Comment by Spicy0903** [comment]
Sounds good but Cryo and water are different Cryo cools more and is more
efficient, info believe mechs are being re worked and the steam
achievements thing I do believe can't be fixed due to steam looking for
that to happen while your in game so if you import a save not much can be
done. Cryofluid does make force projectors last longer. Enemy health bars
exist they flash the more damaged they are same as on your own ship and
last but not least, desync is a massive issue that anuke is aware of and
trying to fix the best you can do currently is to just do /sync when you
notice an anomaly apart from that not much can be done, hopefully our
saviour anuke can find a fix.

**Comment by 1RedOne** [comment]
it'd be cool if some of the ideas that Anuken et all like were made into issues here with 'Help Wanted' or other similar tags.  I wanna help but don't want to step on others efforts :)

**Comment by Spicy0903** [comment]
Anuke doesnt have a help wnated tag because probably doesnt know what he
needs/wants help with

**Comment by oxpa** [comment]
"Destroy enemy core" missions are quite strange: I can beat 'fungal pass' quite quickly (well below 15-th wave) yet I have to defend enemy core not to kill it until I can unlock next location.
May be you could redesign this kind of missions? Say, count victory if everything is destroyed till wave 60 then till wave 30 or whatever looks reasonable.
It's also somehow interesting to play this kind of rogue missions. It may worth investing some time into developing a "real" rogue like mission which you can can't beat with drones or turrets. Say, no flying drones except yours and an isolated area for your base to defend. And you have to destroy a core in another area till waves are too strong.

**Comment by TekhnaeRaav** [comment]
> Enemy health bars should be an option. On maps with multiple bosses, 
it's impossible to tell the health of each boss and damage to individual
 bosses seems to not reflect the total health indicator in expected 
ways.



I agree. This would be very nice.



> Fluid diversity: When water and cryofluid both give the same benefit, 
why would anyone use cryofluid in a building? Make cryofluid more 
powerful in some applications.



Definitely. Cryo-fluid is much more expensive in materials and space than water, it needs an advantage.



> Core resource display on screen at all times. Enable by option maybe?



Yes!



> Molten slag in a scorch turret seems like it ought to be a thing



Yeah,
 maybe it could be more powerful than coal but needs cooling from 
cryo-fluid for a constant stream without overheating, water could work 
for it too but cannot keep a constant stream.



> Force projectors seem to shutdown without reason.



Yeah, they just need a life/overheat meter somewhere.



> I can beat 'fungal pass' quite quickly (well below 15-th wave) yet I 
have to defend enemy core not to kill it until I can unlock next 
location.



These probably should be inverted in the way they handle unlocking levels, so the quicker you beat them the better your score.



---- On Sun, 06 Oct 2019 22:20:43 -0500 ducttapecoder-vt <notifications@github.com> wrote ----


Sorry if any of these are repeats:
 -Fluid diversity: When water and cryofluid both give the same benefit, why would anyone use cryofluid in a building? Make cryofluid more powerful in some applications.
 -Key binding: allow a second key to be bound to an action. I'd like both mousewheel and R to rotate objects for example.
 -Invert filter: allow all blocks except the one you choose to go forwards.
 -Fighter mechs should automatically attack all targets on map when given attack order. Currently, anti-air fighter mechs will head to the enemy spawn point and sit there until an enemy comes close to them. The repair drones fly across the map to get repairs done, why wouldn't your defense drones do the same?
 -When importing data from standalone game into steam game, corresponding achievements should unlock. I had to start a new game data to redo the tutorial, router block, and titanium unlocks. Not a huge deal but annoying.
 -Enable public game option should be a yes/no versus ok/cancel dialogue. I was afraid cancel would cancel the multiplayer game and clicked ok but then my game was open to public.
 -Campaign Story? Are the spores the enemy and have taken control of your own mechs? Are you fighting a different mech army? etc.
 -Diagonal conveyors should wrap around wall blocks rather than just be blocked by them. I spend a ton of time in game getting conveyors to follow walls.
 -Core resource display on screen at all times. Enable by option maybe?
 -Router block contents display by hover. Enable by option?
 -Show junction contents when clicked (or hover)
 -Click and hold resource in block display to hoover. I spend a lot of time cleaning up 'messes' in my router/junction blocks. I usually end up deleting the whole block and placing it back because it's easier than trying to remove the resources from them. If I could click and hold to grab items from a block as they come in, rather than having to continuously click, it would be more viable.
 -Make it more obvious whether deleting a block gives you resources back. My group of friends still doesn't agree whether it loses or returns some/all resources.
 -Make shock tower connections more obvious? I have a really hard time getting the large power towers to work correctly. It always seems like they refuse to connect. Someone usually points out that all the connections are taken, but it's not super obvious to me. Maybe a pop-up or sound to indicate that one of the two blocks has reached connection limit?
 -Make something that sprays environmental effects? There's an achievement for water and electricity on an enemy but getting enemies wet is difficult.
 -Molten slag in a scorch turret seems like it ought to be a thing.
 -Force projectors seem to shutdown without reason. Presumably due to damage or overheat but no way to know. Show a damage indicator? Also, cryofluid should make them last longer.
 -Enemy health bars should be an option. On maps with multiple bosses, it's impossible to tell the health of each boss and damage to individual bosses seems to not reflect the total health indicator in expected ways.
 -On some maps there are ground decorations that look like coal and I end up accidentally mining sand instead, quite annoying. Make them more obviously different.
 -We've had some pretty serious sync, display issues. To one player, a belt will look full, and to another it's half empty. We had one player host and everyone else ended up with pings of over 200000ms due to poor connection and it became difficult to tell what actions on screen were 'real' or 'fake'. Usually we'd determine an enemy was 'taking too long to die' and have to look up at the ping indicator to figure out the problem.

**Comment by IndusRobot** [comment]
I spent a little time trying to find how to change the default server port, but it looks like it might be hardcoded. Having some way to change the port would be really nice, even if it's just a config option tucked into a file somewhere.

**Comment by Tazshelby2016** [comment]
> These probably should be inverted in the way they handle unlocking levels, so the quicker you beat them the better your score.
Unfortunately, this would allow individuals to get a bonus by leaving on the first opportunity.

**Comment by Maxattax97** [comment]
My friends and I were playing PvP today and a "_technique_" we came up with was dumping loads of cheap copper walls *all around* enemy turrets. It distracts them and slows them down tremendously, granting you enough time to place down towers of your own to destroy theirs in a turf war. **It also prevents the enemy from building anything of use** since the walls are blocking all buildable spots. This only happens in no-man's land of PvP maps, since a sizable portion of the map is protected from enemy building.

We were wondering if this is avoidable using some kind of "turf claiming" system. A simple example would be a hook that fires before a block is placed to check that there aren't any buildings in range (the range is small, say about the diameter of a Scorch tower). If not, the block can be placed; otherwise, it's canceled. The other question is whether or not this is desirable/valid gameplay, which is a possibility.

Willing to implement this feature myself and make a PR, just want some input before I do it.

**Comment by Spicy0903** [comment]
Sounds great!

**Comment by CyberFoxHax** [comment]
The long touch delay on mobile needs to be configurable.
It feels like 1000ms is ok for beginners. But as you get seasoned and start building tons of lines it's too slow.
I'd like to see a slider in the settings where i could move it between 100ms - 1000ms.

**Comment by runiArnsbjornson** [comment]
1 -- would love to see more diversity in ennemy (i know it's on trello and part of the future updates but i'd like to share some idea here) 

1.1 - shielding boss : a huge walking force projector that protect smaller swarm and makes them way more dangerous, doesn't do damage on it's own, have a small cooldown time and not that much health (or may be just a regular mech, rather than a boss)

1.2 - burrowing/stealthed enemy that would require special turret to detect/unborrow (tremor and or detector)

1.2 - sniper mech, that would stop once in range of the defense line and start shooting, slowly but doing high damage, slowly wrecking your defense line, making the allied unit more useful on survival maps

1.3 - big boomer, like a huge sized one that would make an explosion bigger (combine it with the shield and i'm already laughing XD) 

1.4 - drop mech : would appear somewhere else than in the bottlenecked spawn point, would make a omnious noise and show where they will land a good 10~30 sec before hand, (and would also promote a bit more unit usefulness) 

1.5 - nano mech : when they die they would release a smoke screen that prevent targetting around them, making defense unable to target a unit inside the nano smoke

1.6 - Magnet boss, would make every metallic ammo change course to target him when close enough, low to middle damage and LOTS of hp 

1.7 - destroyer, very dangerous boss that could litterally walk on your defense, ignore the blocking and just destroy building by walking on them, slow and only and anti unit weapon equiped. 

1.8 - 


2 -- a more useful minimap, possibly showing the whole map instead of just a zone that can be checked with the zoom out ?

3 -- correcting the weird things in the input/ouput section : exemple the cultivator used 25.2 water but have a production time of 2.33 does it means it uses(25.2/2.33) = 10.8~/sec or 25.2/sec compared to other factories it's unclear : oil extrator uses 9 water PER second even though it has a 1.67 production time, so i guessed it outputed 15 oil every 1.67 but may be i'm wrong

3 -- colorblind mode : PLEASE do something about plastinium and surge alloy, i can't see the difference between them unless i zoom in close to max, i would love to see either a change in color or a different shape for every ressource.

4 -- event : meteorites, electric storm, flooding, eruption, earth quake (or planetquake), trigger for a wave and until the wave is destroy or idk 

5 -- bomb that allows you to destroy some walls, would require a lot to build and needed to be triggered manually, would allow interesting mission where you have little place to start with, and an ever growing army waiting on the other side if you take too much time opening, and or allow hidden ressource/making new path of attacks. would indeed require a new type of undestructible-at-all block but liked the idea

6 -- some AND, OR, XOR logical trigger for conveyor/energy. and some light with different color that could be triggered by presence or absence of a ressource and/or energy 

7 -- more trap block, mines, pressure plate and such 

8 -- and upgrad unit factory that would require huge amount of ressource changing with every level of upgrade and with a very long research time (like ~300 sec) could also be used by ennemy 

9 -- more turrets : 

9.1 - jet cutter turret : slag/water/cryofluid + energy ; makes a very high pressurized jet of slag, capable of dealing damage and {burn, slow, freeze} enemy 

9.2 - big one : makes a huge explosion, which could damage your own building and unit, needs to be manually triggered 

9.3 - nano smoke turret : use huge amount of ressource to create a smoke that cover the surrounding building in a smoke that protect them for a few second from being targeted.

9.4 - disruptive turret : takes control of an enemy mech every X sec, needs energy

9.5 - trio : the same as our good for early duo, but fire in three direction, somehow like a badly made shotgun

9.6 - tesla storm : like the arc but all around and for more energy and far stronger effect, also chain on enemy, the bigger they are the further they chain. 

9.7 - plasma turret : would require cryofluid, and some sort of gas ofc, idk which one or how :x anyway would launch a big explosive and slowing ball that would damage and push enemy, would do moderate to high damage, plus a damage over time part

10 -- differenciate between the effect of boosters in description and make more obvious the different benefices, like water will do speed x1.5, cryofluid x2 a bit like the ammo description does




For those that reached here, thank you, hope you'll like some of my idea and not burn me for them
<3

**Comment by Tazshelby2016** [comment]
> My friends and I were playing PvP today and a "_technique_" we came up with was dumping loads of cheap copper walls _all around_ enemy turrets. It distracts them and slows them down tremendously, granting you enough time to place down towers of your own to destroy theirs in a turf war. **It also prevents the enemy from building anything of use** since the walls are blocking all buildable spots. This only happens in no-man's land of PvP maps, since a sizable portion of the map is protected from enemy building via distance from the core.
> 
> We were wondering if this is avoidable using some kind of "turf claiming" system. A simple example would be a hook that fires before a block is placed to check that there aren't any buildings in range (the range is small, say about the diameter of a Scorch tower). If not, the block can be placed; otherwise, it's canceled. The other question is whether or not this is desirable/valid gameplay.
> 
> Willing to implement this feature myself and make a PR, just want some input before I do it.

They might just want to ignore walls altogether. The walls can be modified to help turrets, and they can just shoot over them.

**Comment by matthias4217** [comment]
My little suggestion today would be to **add server commands**. In fact, in the singleplayer client, it's possible to become a host, and then there are several parameters available, such as the way waves are launched. However, these options don't exist on the server. I'm not sure how much work it would be, and I know that there might be mods to configure this, but I think it would be great to have a server with many gamerules to tweak in order to customize the experience of players.

(Anyway, I really love the game !)

**Comment by 1RedOne** [comment]
Enemy Base Under Attack notice

Can't tell you how many times my units work their way through the enemy defenses and take down the base...and I don't realize it because I was busy optimizing or reruning a supply line for better aesthetics.  Would be cool for the miminap to indicate that the enemy base is being actively attacked somehow.

**Comment by Malacosa** [comment]
> Enemy Base Under Attack notice
> 
> Can't tell you how many times my units work their way through the enemy defenses and take down the base...and I don't realize it because I was busy optimizing or reruning a supply line for better aesthetics. Would be cool for the miminap to indicate that the enemy base is being actively attacked somehow.

Sounds like what you need to do is have a command center and tell your units to rally/retreat OR use doors to prevent your units from attacking until you want to end the map.  I personally find doors really useful, but my one suggesting here is that the enemy AI should consider doors to be open space, as using them definitely changes the AI pathing.

**Comment by 1RedOne** [comment]
I'd love to be able set attack targets for my units, on that topic...

**Comment by RumbleballTheReal** [comment]
Currently you can only have one wave at a time. Would be nice to just spam the next wave button to progress faster. Maybe only allow this behavior until a Launch-Wave appears.

**Comment by rakesh491** [comment]
Hi, I have been playing mindustry for a month now. The most fun part of the game is in building new structures and hence giving a new feature to the base. However the frequent enemy waves makes it very difficult to focus on the base and more on the defences. I wished that if the time duration between two waves was more then players can focus on building the base too. Also gradually increase he enemy strengths.

**Comment by willis936** [comment]
iOS has achievement notifications when they’re unlocked but as far as I can tell there is no way to view the achievement list.  Does this hook into iOS’s game center?

**Comment by willis936** [comment]
Industrial games like this have a lot of overlap with reactor idler games (industrialCraft2’s nuclear reactor and its derivatives, such as reactor-incremental and reactor-idle.  These games have fundamental issues related to rewarding staring at the screen.  I think the design of Mindustry brilliantly gets around this issue and is a strong framework for adding in reactor idler game components.  There is coolant and generators, etc.  I think going further in this direction would be cool.  Like multiblock structures involving heat sources, exchangers, and vents, with the typical tradeoffs (heat cells that are adjacent to other heat cells are more efficient, but more difficult to cool).  This would be for midgame power generation.  Conceptually, I really like the impulse reactor for endgame power production.

**Comment by Arkanic** [comment]
@rakesh491 you can easily do this yourself, as you can customise waves in the map editor.

**Comment by TheMeltdown201** [comment]
New feature idea : wind turbines

Power generation : 30/sec
Build Cost : you see.
Health : same as combustion generator.

2: Stirling Generator

Power Generation : 180/sec
Requirements : Slag and Water

3: Plastanium/Lead/Graphite/Silicon Walls.

**Comment by joshuaptfan** [comment]
# Hotkeys suggestions for building

[Reddit post](https://www.reddit.com/r/Mindustry/comments/dlmklh/hotkeys_suggestions_for_building/)

Hotkeys would allow players to select blocks without moving their mouse away from what they're focusing on. Most RTS's implement hotkeys for building and unit actions.

These suggestions are not alternatives to each other, ideally they would all be implemented to allow for a range of input styles.

## Number keys
This scheme is deal for keyboard.

This idea is to have the first press of a number key switch block tab, and second press selects a specific block.

- You can just press once to switch tabs if you haven't memorized block numbers yet. ("3" switches to distribution.)
- Numbers should correspond to tabs left-to-right, top-to-bottom. ("1" defense, "2" production, "3" distribution...)
- Some tabs have more than 10 blocks, e.g. crafting tab (graphite press, etc.). For those, to select the 12th block, you would press "7,1,2" in rapid succession. "7,0" and "7,1,0" would select the same block ("0" is a shortcut for 10).
- Your hotkey combo should expire after a short time like 500ms if you didn't finish it, so that a combo can be easily canceled without requiring additional key presses. If you press "7", pause, "1", it should switch to crafting then defense tab. (Specifically, you should have 500ms to press the next key in the combo, e.g. 500ms between each press of a 3 key combo.) Perhaps this timeout should be configurable in settings.
- Sandbox has extra blocks that would screw with muscle memory (e.g. conveyor would be on "3,3" in sandbox). I suggest moving sandbox blocks to the end of the list.
- These keys should be rebindable as building hotkeys 1-10.

## Hotbar keybinds
Alternatively or additionally, hotkeys would work well for quickly selecting blocks in the hotbar, [whenever that is implemented.](https://trello.com/b/aE2tcUwF/mindustry-trello)

## Cycling selection
I tested mapping keys/mouse to my controller on Steam, and it was very playable, but moving the cursor to the corner to select blocks was painfully slow.

This scheme uses fewer keys, and is ideal for controllers. Console support is planned on the Trello, so controller support is relevant. Keybinds should still be made available on desktop so PC users can map controller buttons to them.

Keys: left/right tab cycle, arrow keys to choose block, "select" keybind to select block.

Controller example: LB/RB tab cycle, D-pad to choose block, A to select block.

- GUI: Display a box around the currently chosen block, only if tab cycle/arrow keys are pressed.
- The chosen block should be remembered for each tab individually, so that it will be easy to reuse a block from a tab after switching away.

Also, for better controller support, hotkeys besides Esc should work in the main menu. (Arrow keys, Enter)

**Comment by TekhnaeRaav** [comment]
>Plastanium/Lead/Graphite/Silicon Walls.


For Silicon walls, and maybe Plastanium as a higher grade version, you could give them a special Absorbtion ability, for example: Only when the walls recieve X damage points within 2-seconds will they actually recieve the damage. In this case they would be effective walls for areas that don't get heavily hit as you wouldn't need healers as much. And maybe give them heat/flamethrower resistance due to silicon's natural heat resistance. For balancing's sake, these walls should have less health than their rarity would suggest too.

---- On Tue, 22 Oct 2019 10:45:43 -0500 TheMeltdown201 <mailto:notifications@github.com> wrote ----


New feature idea : wind turbines

Power generation : 30/sec
 Build Cost : you see.
 Health : same as combustion generator.
2: Stirling Generator

Power Generation : 180/sec
 Requirements : Slag and Water
3: Plastanium/Lead/Graphite/Silicon Walls.

**Comment by majora2007** [comment]
Heck of a lot of ideas here, here are a few I didn't see.

The top right mini-map is kinda pointless. It doesn't show active enemies, you can't use it to quick jump to an area on the map. I like the idea of showing you overall "circuit", but I don't think that should be there. I recommend:
- Minimap to jump camera, show active bosses or moving enemies.
- Add a print map button which generates a big jpg/png of your map without AI units.

Machines really need to show how much power they get or how much they need. The progress bar works, but it's not that well understood.

The power situation is kinda confusing. It is confusing when you are low on power or not. The animation is not clear to me. A notification system might be useful for things like this or even if it is triggered in game from a battery being depleted, send a notification.

It would be extremely cool if you could build routes for your mechs to walk if no enemies are present or ability to select a group of units at a time and tell them to attack an area. Units currently are all the same (I believe there is a Trello card on this) and there is no indication on their health or damage. 

Overall, amazing game. Balancing is really good.

**Comment by TenguOnTheTop** [comment]
A small thing, but add the ability to see the 'blast range' on a Thorium Generator, so that one can gauge what's at risk of being destroyed in case of an overheat.

**Comment by Luxray5474** [comment]
> The top right mini-map is kinda pointless. It doesn't show active enemies, you can't use it to quick jump to an area on the map. 

You have a point; the minimap doesn't serve much of a purpose, especially since you and other units are pretty thick squares. It would be nice to have a detailed view but that would undoubtedly require more system resources.

> Machines really need to show how much power they get or how much they need. The progress bar works, but it's not that well understood.

Specifically how? It's been pretty understandable for me, and I have not seen anyone having a problem with it. [current power] / [required power] is pretty simple.

> The power situation is kinda confusing. It is confusing when you are low on power or not. The animation is not clear to me. A notification system might be useful for things like this or even if it is triggered in game from a battery being depleted, send a notification.

Notifications are a great idea, but really if you have a well-designed base, you will have at least 4 nodes on your screen at a time 75% of the time. I think a resource view comeback would be great; remember we had those in the first versions of the game above the block menu? A lot of major sandbox/td/sim games have those like it's a standard. It would be nice to have it back, not just with items. but also with power, health, and number of units, in a production-usage style, like `⚡ +107.55 -67.88`, the production being green, and usage being red, and production becoming orange one the usage reaches the limit. 

> It would be extremely cool if you could build routes for your mechs to walk if no enemies are present or ability to select a group of units at a time and tell them to attack an area. Units currently are all the same (I believe there is a Trello card on this) and there is no indication on their health or damage.

The first part, awesome idea. That would create flanking strats and more, and make the game also include RTS as a playstyle. And healthbars on units were awesome, please bring them back anuke.

> Also I'm not sure, but do doors have the same "health" rating as copper walls?

You can read up on [door info here](https://mindustrygame.github.io/wiki/blocks/defense/door/). [We have a wiki if you haven't heard yet!](https://mindustrygame.github.io/wiki/)

**Comment by majora2007** [comment]
@Luxray5474 I think another good idea would be having a link to the Wiki in-game. I'm personally not a fan of jumping out to google to research things, esp with games like this. You want people to drive new solutions, not just stumble upon a super efficient solution online.

Also, idk if I mentioned it, but mechs should be able to open doors without me.

**Comment by Luxray5474** [comment]
I agree. I Might start a pull request for that soon.

**Comment by 1RedOne** [comment]
For alerts, I want my turrets to alert me if they are out of ammo too!

**Comment by TekhnaeRaav** [comment]
Um, There is a link to the wiki in the "About" page in Mindustry.



---- On Thu, 24 Oct 2019 10:50:37 -0500 L5474 <notifications@github.com> wrote ----


I agree. I Might start a pull request for that soon.

**Comment by knerf999** [comment]
Hey Anuken & Arkanin,

I was just wondering about something.
I threw some numbers at the wall and came to the conclusion that Airblast drills are not that much of an improvement (if at all) over laser drills.
Though they are advertised as an "ultimate drill" they have little benefit in my play experience.

First off
The base drill speeds are equal.
They both sport a 0.21 yield per occupied square of mine-able  resource.

the lazer drill uses less energy per mineable square (66e/9sq=) 7.33e/sq
where the airblast drill consumes (180e/16sq=) 11.25e/sq

the only benefit the airblast drill seems to have is 26% reduced water consumption
(6w/16sq=)0.375 for the airblast per square
(4.6w/9sq=)0.511 for the lazer drill.

Even buildtime isnt a factor here.
The airblast drill takes 0.22 seconds per square to build
And the lazerdrill a mere 0.16 seconds per square.

The size of the thing hardly ever covers enough resources to justify building one, so it is often outmined by placing laserdrills instead.

On top of that, every airblast drill requires about as much rescources as two laser drills. 
Other than using more coal for the laser drill versus using thorium for the airblast drill.

I could just be missing something obvious, looking at numbers like this is work for statisticians and other devil worshippers afterall.

**Comment by Shawak** [comment]
Use a random map on game start if you don't specify a map instead of not starting at all.

**Comment by TechFish3000** [comment]
Allow mod to check other mods for items, so they can use the same and/or create dependencies. I have recently tried to make a mod but it failed loading items from the other due to this not being a thing. Until this is fixed, my mods will have to wait.

**Comment by TechFish3000** [comment]
#894

**Comment by TekhnaeRaav** [comment]
> Airblast drills are not that much of an improvement (if at all) over laser drills.



@knerf999 I agree with some of your points here, two laser drills tend to work just as well as a single airblast drill. The airblast drills are still very useful as they give you more variability in your use of space and using less water is also somewhat significant in areas lacking sufficient water sources because of their need for only one water extractor, but there really is no difference between them. We should keep in mind, however, the fact that it may overpower the player to have anything more efficient, I would like to think of it as simply a higher progression point for drills.



I would propose the allowance of Airblast drills to take Cryofluid to be more efficient. Also doesn't it seem like mechanical and pneumatic drills should take oil?





---- On Fri, 25 Oct 2019 09:58:10 -0500 knerf999 <notifications@github.com> wrote ----



Hey Anuken & Arkanin,

I was just wondering about something.
 I threw some numbers at the wall and came to the conclusion that Airblast drills are not that much of an improvement (if at all) over laser drills.
 Though they are advertised as an "ultimate drill" they have little benefit in my play experience.
First off
 The base drill speeds are equal.
 They both sport a 0.21 yield per occupied square of mine-able  resource.
the lazer drill uses less energy per mineable square (66e/9sq=) 7.33e/sq
 where the airblast drill consumes (180e/16sq=) 11.25e/sq
the only benefit the airblast drill seems to have is 26% reduced water consumption
 (6w/16sq=)0.375 for the airblast per square
 (4.6w/9sq=)0.511 for the lazer drill.
Even buildtime isnt a factor here.
 The airblast drill takes 0.22 seconds per square to build
 And the lazerdrill a mere 0.16 seconds per square.
The size of the thing hardly ever covers enough resources to justify building one, so it is often outmined by placing laserdrills instead.

On top of that, every airblast drill requires about as much rescources as two laser drills.
 Other than using more coal for the laser drill versus using thorium for the airblast drill.
I could just be missing something obvious, looking at numbers like this is work for statisticians and other devil worshippers afterall.

**Comment by willis936** [comment]
The visual style/design seems ripe for vector art.  svg can be animated and through some kind of wizardry be interactive.  This would allow for crisp visuals regardless of render resolution.  However it would almost certainly be lower performance and a serious pain to implement.

**Comment by Arkanic** [comment]
@knerf999  I agree, airblasts don't give much of an advantage
~~also i've barely contributed anything to this game~~

**Comment by Jaeiya** [comment]
I would like to see a page dedicated to feature requests, where it's easier for us to see and upvote the good ideas. It's too hard on github to have to plow through all the comments which have a conglomeration of random ideas in a single entry.

**Comment by zicklag** [comment]
I agree that it would be nice to have dedicated pages to features so that they could be discussed and voted on without having to have them all intermingled with each-other.

**Comment by majora2007** [comment]
@Anuken Could you create a feathub so these ideas can be put on and voted against.

**Comment by Anuken** [comment]
Yeah, I'll be moving to [FeatHub](https://feathub.com/Anuken/Mindustry) now, good idea. A badge is now in README.

**Comment by TheMeltdown201** [comment]
Hello anuke, i just thought of an idea (Custom Rules), the idea consists to custom crafting (Choosing matérial and it's cost for a spécific block). Thanks you if you read this.

**Comment by Pointifix** [comment]
For the hexed pvp mode it would be great if build protection zones could be placed independent of cores, or just being able to change the sphere protection to a square protection zone around the cores. This would be usefull to close gaps between the cores in hexed pvp mode (with removed Hexagons ofc.)

For example it would be desireable to change the protection shape of cores in the Rules object:

in addition to this one: `rules.enemyCoreBuildRadius` 

for instance: `rules.enemyCoreBuildShape`

![image](https://user-images.githubusercontent.com/35230128/76271141-d3aa3900-6277-11ea-9183-7f36e892af01.png)

**Comment by Galactic-UN** [comment]
Hmmm, maybe can you add water/material junction to mindustry?
